build_runner/lib/src/server/hot_reload_client.dart

```diff
@@ -2,14 +2,107 @@
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
 
+@JS()
+library hot_reload_client;
+
+import 'dart:async';
+import 'dart:convert';
 import 'dart:html';
 
+import 'package:js/js.dart';
+import 'package:js/js_util.dart';
+
+final _assetsDigestPath = r'$assetDigests';
 final _buildUpdatesProtocol = r'$livereload';
 
-main() {
+@JS('Map')
+abstract class JsMap<K, V> {
+  @JS()
+  external Object keys();
+
+  @JS()
+  external V get(K key);
+}
+
+@anonymous
+@JS()
+class DartLoader {
+  @JS()
+  external JsMap<String, String> get urlToModuleId;
+
+  @JS()
+  external void forceLoadModule(
+      String moduleId, void Function(Object module) callback);
+}
+
+@JS(r'$dartLoader')
+external DartLoader get dartLoader;
+
+@JS('Array.from')
+external List jsArrayFrom(Object any);
+
+List<String> get moduleUrls {
+  return List.from(jsArrayFrom(dartLoader.urlToModuleId.keys()));
+}
+
+String moduleIdByPath(String path) =>
+    dartLoader.urlToModuleId.get(window.location.origin + '/' + path);
+
+Future<Object> reloadModule(String moduleId) {
+  var completer = Completer<Object>();
+  dartLoader.forceLoadModule(moduleId, allowInterop(completer.complete));
+  return completer.future;
+}
+
+class ReloadHandler {
+  final String Function(String) _moduleIdByPath;
+  final FutureOr<Object> Function(String) _reloadModule;
```

Comment:
Should this be a typedef to make it easier to mock? I find it a little odd in the test to have an abstract class of methods.

Reply:
I haven't find any examples how to mock functions/typedefs directly outside of a class/interface. Also packing functions into class allows to use `verifyZeroInteractions` and `verifyNoMoreInteractions`